airbyte-server/src/main/java/io/airbyte/server/converters/SchemaConverter.java

```diff
@@ -35,71 +35,93 @@
-                          .map(
-                              apiField -> {
-                                final Field field = new Field()
-                                    .withName(apiField.getName())
-                                    .withSelected(apiField.getSelected());
-                                // todo (cgardens) - this is a hack to handle complex types (e.g. anyOf). they will not have a type
-                                // field set, so we just call them "object". we need to revisit this ASAP.
-                                if (apiField.getDataType() != null) {
-                                  field.withDataType(Enums.convertTo(apiField.getDataType(), DataType.class));
-                                } else {
-                                  field.withDataType(DataType.OBJECT);
-                                }
-                                return field;
-                              })
-                          .collect(Collectors.toList());
+        sourceSchema.getStreams()
+            .stream()
+            .map(apiStream -> {
+              final List<Field> persistenceFields = apiStream.getFields()
+                  .stream()
+                  .map(SchemaConverter::toPersistenceField)
+                  .collect(Collectors.toList());
 
-                  return new Stream()
-                      .withName(apiStream.getName())
-                      .withFields(persistenceFields)
-                      .withSelected(persistenceFields.stream().anyMatch(Field::getSelected));
-                })
+              return new Stream()
+                  // immutable
+                  // todo (cgardens) - not great that we just trust the API to not mutate these.
+                  .withName(apiStream.getName())
+                  .withFields(persistenceFields)
+                  .withSupportedSyncModes(apiStream.getSupportedSyncModes()
+                      .stream()
+                      .map(e -> Enums.convertTo(e, io.airbyte.config.StandardSync.SyncMode.class))
+                      .collect(Collectors.toList()))
+                  .withSourceDefinedCursor(apiStream.getSourceDefinedCursor())
+                  .withDefaultCursorField(apiStream.getDefaultCursorField())
+                  // configurable
+                  .withSyncMode(Enums.convertTo(apiStream.getSyncMode(), io.airbyte.config.StandardSync.SyncMode.class))
+                  .withCursorField(apiStream.getCursorField())
+                  .withSelected(persistenceFields.stream().anyMatch(Field::getSelected));
+            })
             .collect(Collectors.toList());
 
     return new Schema().withStreams(persistenceStreams);
   }
 
+  private static Field toPersistenceField(SourceSchemaField apiField) {
+    final Field field = new Field()
+        .withName(apiField.getName())
+        .withSelected(apiField.getSelected());
+    // todo (cgardens) - this is a hack to handle complex types (e.g. anyOf). they will not have a type
```

Comment:
Do you have a sense of what the proper way to do it is? 

How risky is it that this hack lives for too long? (around other part of the code making assumptions about it and starting to build upon this hack)

Reply:
> Do you have a sense of what the proper way to do it is?

yeah, i think what we will want to do is treat the schema fully as json schema instead of forcing it into a table / column configuration. 

> How risky is it that this hack lives for too long? (around other part of the code making assumptions about it and starting to build upon this hack)

It's something I'd like us to tackle after incremental. It has been nice that as we have iterated on the internal catalog, we have not needed to worry about breaking the UI or API. This has allowed us to move towards what needed much faster. That being said, I think the basic structure of the catalog should stop changing after incremental and it should be safer to make the UI / API configuration as expressive as what we have internally. 

I'd probably say, if we got to the end of January without having fixed this, we've made a mistake. That being said, I think it will be a medium difficulty UI project (just handling nested fields and the full feature set of jsonschema) and relatively trivial change in the API, so I'm not too worried about us being able to get it done.